scalableminds/webknossos
# Make merger mode work for all skeleton modifications

Merger mode only worked for basic node/tree creation/deletion cases (see #4655 for an explanation), but didn't work for undo/redo or when merging/splitting trees.

By switching from overwriting specific redux actions to acting on the result of our tree differ, all of these cases are now handled. The resulting code is even easier as the merger mode only needs to react to two diffing actions: `createNode` and `deleteNode`. Complex redux actions like mergeTree, result in low-level diffing actions:
- `deleteNode` actions for each node of the tree that was deleted as part of the merge
- `createNode` actions for all of the same nodes, but in the tree that "survived" the merge
- a `deleteTree` action that merger mode doesn't care about, similar an `updateTree` action which only modifies tree properties
Similar for undo/redo/tree merging/splitting. This is actually an advantage of the "primitive" low-level diffing we do.

I fixed a couple of other bugs while I was at it (not all of them have issues):
- Progress notifications are no longer displayed if another mapping was activated before (#4514)
- The ID Mapping toggle in the segmentation tab was buggy and remained disabled even though a mapping was activated programmatically or by selecting a value from the mapping dropdown
- The merger mode mapping was not deactivated when deactivating merger mode and remained applied
- There was a race condition in merger mode tasks where sometimes a texture wasn't correctly initialized, resulting in all segments being red
- Not a bug, but I replaced the "You set a node too close to gray ..." modal with a Toast, as was wished for in this Discuss post: https://discuss.webknossos.org/t/remove-popup-from-merger-mode-script/1052.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a tracing with a segmentation and activate merger mode. Execute all kinds of tree-modifying actions - the merger mode mapping should remain consistent. (Undo, Redo, Node Deletion, Tree Split, Tree Merge, ...)
- Activate another mapping, then activate merger mode and set nodes. No progress notifications should appear.
- Deactivate Merger Mode. The mapping should no longer be applied.
- Activate a mapping through the API: `webknossos.apiReady(3).then(api => api.data.setMapping("segmentation", {1: 2, 3: 4}))`. The ID Mapping toggle in the segmentation tab should become activated.

### Issues:
- fixes #4655
- fixes #4514 

------
